### PR TITLE
Set the most logical config defaults for the wizard

### LINKED ISSUE #181: Types emitted on OData client should implement INotifyPropertyChanged interface
### Description
OData client uses `DataServiceCollection<T>` to support property tracking.

#### Example:
```csharp
DataServiceCollection<Order> orders = new DataServiceCollection<Order>(dataServiceContext.Orders.Where(d => d.Id == 1));
// Update Amount
orders[0].Amount = 37;
dataServiceContext.SaveChanges();
```
The payload will comprise only of the updated properties and will be sent as a PATCH request:
```
{
    "@odata.type": "#NS.Models.Order",
    "Amount": 37
}
```

However, to use `DataServiceCollection<T>`, the client type needs to implement `INotifyPropertyChanged` interface. The `Order` type for example would look as follows:
```csharp
[Key("Id")]
public partial class Order : BaseEntityType, global::System.ComponentModel.INotifyPropertyChanged
{
	public static Order CreateOrder(int ID, DateTimeOffset orderDate, decimal amount)
	{
		Order order = new Order();
		order.Id = ID;
		order.OrderDate = orderDate;
		order.Amount = amount;
		return order;
	}

	public virtual int Id
	{
		get
		{
			return this._Id;
		}
		set
		{
			this.OnIdChanging(value);
			this._Id = value;
			this.OnIdChanged();
			this.OnPropertyChanged("Id");
		}
	}
	private int _Id;
	partial void OnIdChanging(int value);
	partial void OnIdChanged();

	public virtual DateTimeOffset OrderDate
	{
		get
		{
			return this._OrderDate;
		}
		set
		{
			this.OnOrderDateChanging(value);
			this._OrderDate = value;
			this.OnOrderDateChanged();
			this.OnPropertyChanged("OrderDate");
		}
	}
	private DateTimeOffset _OrderDate;
	partial void OnOrderDateChanging(DateTimeOffset value);
	partial void OnOrderDateChanged();

	public virtual decimal Amount
	{
		get
		{
			return this._Amount;
		}
		set
		{
			this.OnAmountChanging(value);
			this._Amount = value;
			this.OnAmountChanged();
			this.OnPropertyChanged("Amount");
		}
	}
	private decimal _Amount;
	partial void OnAmountChanging(decimal value);
	partial void OnAmountChanged();
	/// <summary>
	/// This event is raised when the value of the property is changed
	/// </summary
	[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.OData.Client.Design.T4", "#VersionNumber#")]
	public event global::System.ComponentModel.PropertyChangedEventHandler PropertyChanged;
	/// <summary>
	/// The value of the property is changed
	/// </summary>
	/// <param name="propertyName">Property name</param>
	[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.OData.Client.Design.T4", "#VersionNumber#")]
	protected virtual void OnPropertyChanged(string propertyName)
	{
		if (this.PropertyChanged != null)
		{
			this.PropertyChanged(this, new global::System.ComponentModel.PropertyChangedEventArgs(propertyName));
		}
	}
}
```
**Note:** The class has been stripped to make it less verbose with emphasis being placed on the fragments that support property tracking. Each property setter calls the `OnPropertyChanged(string)` method.

Currently, the client types emitted by OData Connected Service extension do not implement the `INotifyPropertyChanged` interface. An exception is thrown if one tries to use `DataServiceCollection<T>` in OData client.

> System.InvalidOperationException: 'An attempt to track an entity or complex type failed because the entity or complex type 'NS.Models.Order' does not implement the INotifyPropertyChanged interface.'

This in turn makes it impossible to support property tracking and alternatives available result into the whole entity being sent over the wire instead of only the updated properties.

#### Additional context
The CodeGen tool used to emit client types that implement `INotifyPropertyChanged` interface. It'd seem that when a decision was made to discontinue active development of CodeGen, this capability was not migrated to OData Connected Service.

### PR DESCRIPTION
**Pull request by gathogojr**
This pull request fixes #181 

Set the most logical config defaults for the wizard. Below is a detailed explanation of the changes:

1. Restore OCS wizard default settings that were inadvertently dropped with the introduction of the settings persistence feature, i.e.
   - GeneratedFileNamePrefix => `Reference`
   - NamespacePrefix => `Reference`
   - UseDataServiceCollection => `true`
   - EnableNamingAlias => `true`
   - IgnoreUnexpectedElementsAndAttributes => `true`
   - ServiceName => `OData Service` 

2. ~Fix handling of **most recently used** (MRU) endpoints. Presently, the `UserSettings` class exposes MRU endpoints as an `ObservableCollection` that one is able to manipulate directly (add/remove/etc) and violate the whole idea of "most recently used". This is fixed by changing `MruEndpoints` property type to `ReadOnlyObservableCollection` and providing a public method `AddMruEndpoint` through which the MRU endpoints collection can be manipulated~

3. Fix binding of user settings to the wizard pages. There was a lot of unnecessary duplication of code to support binding with lots of copying logic in multiple places. This is fixed by having all the wizard pages bind directly to the properties in a `UserSettings` instance.

4. Tests would change the settings file that is used when creating Connected Services from Visual Studio. This is fixed by making it possible to create a "named" user settings config file in test scenarios

5. Refactor to make the code leaner and cleaner

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=187) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=187) before we can accept your contribution.<br/><br/>:x: gathogojr [sign now](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=187)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OData/ODataConnectedService?pullRequest=187) it.</sub>

File: src/Common/UserSettingsPersistenceHelper.cs [hunk 1]
```diff
@@ -112,7 +113,7 @@ public static void Save(object userSettings, string providerId, string name, Act
 
         private static string GetStorageFileName(string providerId, string name)
         {
-            return providerId + "_" + name + ".xml";
+            return providerId + "." + name + ".xml";
```
Comment (habbes) [inline src/Common/UserSettingsPersistenceHelper.cs:116]: What was the purpose of changing from `_` to `.`  in the filename?

File: src/Models/UserSettings.cs [hunk 2]
```diff
@@ -16,13 +17,24 @@ namespace Microsoft.OData.ConnectedService.Models
...
-        private ConnectedServiceLogger logger;
+        private readonly ConnectedServiceLogger logger;
 
-        [DataMember]
-        public ObservableCollection<string> MruEndpoints { get; set; }
+        private readonly string configName;
+
+        private Action<UserSettings> onUserSettingsLoaded = (userSettings) =>
+        {
+            userSettings.MruEndpoints = new ReadOnlyObservableCollection<string>(userSettings.mruEndpoints);
```
Comment (habbes) [inline src/Models/UserSettings.cs:28]: How come here you are not checking whether `userSettings != null` but in other cases you in this file you are?

File: src/Models/UserSettings.cs [hunk 3]
```diff
@@ -93,59 +105,91 @@ internal class UserSettings
...
+                {
+                    // onLoaded action is currently triggered only if settings are loaded from config file
+                    // Do this defensively all the same...
+                    if (userSettings != null)
+                    {
+                        this.mruEndpoints = new ObservableCollection<string>(userSettings.mruEndpoints);
+                        this.MruEndpoints = new ReadOnlyObservableCollection<string>(this.mruEndpoints);
+                    }
+                },
+                logger: logger);
```
Comment (habbes) [inline src/Models/UserSettings.cs:137]: If I understand correctly, this automatically attempts to load the settings from file when you instantiate `new UserSettings()` even without calling `userSettings.Load()`?

**Review by habbes — APPROVED** [review]
Thanks for cleaning things up.

File: src/Models/UserSettings.cs [hunk 4]
```diff
@@ -72,10 +74,10 @@ internal class UserSettings
         [DataMember]
         public bool IncludeWebProxyNetworkCredentials { get; set; }
 
-        [DataMember]
+        [IgnoreDataMember]
```
Comment (KenitoInc) [inline src/Models/UserSettings.cs:77]: Also add `[IgnoreDataMember]` to `CustomHttpHeaders` since we my have authentication tokens

File: src/Models/UserSettings.cs [hunk 5]
```diff
@@ -93,59 +95,89 @@ internal class UserSettings
...
-            this.MruEndpoints = new ObservableCollection<string>();
+            this.configName = configName;
+            this.logger = logger;
+            // Desired defaults
+            GeneratedFileNamePrefix = Constants.DefaultReferenceFileName;
+            NamespacePrefix = Constants.DefaultReferenceFileName;
+            ServiceName = Constants.DefaultServiceName;
+            UseDataServiceCollection = true; // To support entity and property tracking
+            EnableNamingAlias = true; // To force upper camel case in the event that entities are named in lower camel case
+            IgnoreUnexpectedElementsAndAttributes = true;
```
Comment (KenitoInc) [inline src/Models/UserSettings.cs:108]: Maybe you add a comment similar to the way you did for `UseDataServiceCollection` and `EnableNamingAlias`

File: src/Models/UserSettings.cs [hunk 6]
```diff
@@ -16,10 +17,11 @@ namespace Microsoft.OData.ConnectedService.Models
...
-        private const string Name = "Settings";
         private const int MaxMruEntries = 10;
 
-        private ConnectedServiceLogger logger;
+        private readonly ConnectedServiceLogger logger;
+
+        private readonly string configName;
 
         [DataMember]
         public ObservableCollection<string> MruEndpoints { get; set; }
```
Comment (KenitoInc) [inline src/Models/UserSettings.cs:27]: Since we have a public `get` and `set` here, I think `MruEndPoints` can be manipulated wrongly here.

File: src/ODataConnectedServiceWizard.cs [hunk 7]
```diff
@@ -32,90 +32,87 @@ internal class ODataConnectedServiceWizard : ConnectedServiceWizard
...
-            if (this.Context != null && this.Context.IsUpdating)
+
+            if (Context != null && Context.IsUpdating)
             {
-                ConfigODataEndpointViewModel.Endpoint = this._serviceConfig?.Endpoint;
-                ConfigODataEndpointViewModel.EdmxVersion = this._serviceConfig?.EdmxVersion;
-                ConfigODataEndpointViewModel.ServiceName = this._serviceConfig?.ServiceName;
-                ConfigODataEndpointViewModel.CustomHttpHeaders = this._serviceConfig?.CustomHttpHeaders;
+                LoadUserSettingsFromServiceConfig();
+                ConfigODataEndpointViewModel.EdmxVersion = ServiceConfig?.EdmxVersion;
```
Comment (KenitoInc) [inline src/ODataConnectedServiceWizard.cs:70]: Why did we stop adding the `Endpoint` and `ServiceName` to the View model?

**Review by KenitoInc — CHANGES_REQUESTED** [review]
A few comments and questions!

File: src/Common/UserSettingsPersistenceHelper.cs [hunk 8]
```diff
@@ -36,7 +36,8 @@ public static void Save(object userSettings, string providerId, string name, Act
                         {
                             // note: this overwrites existing settings file if it exists
                             stream = file.OpenFile(fileName, FileMode.Create);
-                            using (var writer = XmlWriter.Create(stream))
+
+                            using (var writer = XmlWriter.Create(stream, new XmlWriterSettings { Indent = true }))
```
Comment (marabooy) [inline src/Common/UserSettingsPersistenceHelper.cs:40]: Do we need to indent this or can this remain the same as the file is not usually read by human?

File: src/Models/UserSettings.cs [hunk 9]
```diff
@@ -93,59 +95,92 @@ internal class UserSettings
...
+                        foreach(var endpoint in userSettings.MruEndpoints)
+                        {
+                            this.MruEndpoints.Add(endpoint);
+                        }
+                    }
+                },
+                logger: logger);
         }
 
+        public UserSettings(string configName = "Settings")
```
Comment (marabooy) [inline src/Models/UserSettings.cs:131]: Docs for the ctor.

File: src/ODataConnectedServiceWizard.cs [hunk 10]
```diff
@@ -32,90 +32,88 @@ internal class ODataConnectedServiceWizard : ConnectedServiceWizard
...
                 // Restore the advanced settings to UI elements.
                 AdvancedSettingsViewModel.PageEntering += AdvancedSettingsViewModel_PageEntering;
             }
 
-            this.Pages.Add(ConfigODataEndpointViewModel);
-            this.Pages.Add(SchemaTypesViewModel);
-            this.Pages.Add(OperationImportsViewModel);
-            this.Pages.Add(AdvancedSettingsViewModel);
-            this.IsFinishEnabled = true;
+            Pages.Add(ConfigODataEndpointViewModel);
```
Comment (marabooy) [inline src/ODataConnectedServiceWizard.cs:81]: To this. or not to?

File: src/Common/ExtensionMethods.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.OData.ConnectedService.Common
+{
+    internal static class ExtensionMethods
+    {
+        public static void CopyPropertiesFrom(this object target, object source)
```
[the remark this answers is not in the document]
Comment (gathogojr) [reply src/Common/ExtensionMethods.cs:21]: Done

File: src/Common/UserSettingsPersistenceHelper.cs (hunk 8 above)
Comment (gathogojr) [reply src/Common/UserSettingsPersistenceHelper.cs:40]: During development when you may need to verify the content, it's extremely hard to follow. I reasoned there would be no harm in having it properly formatted

File: src/Common/UserSettingsPersistenceHelper.cs (hunk 1 above)
Comment (gathogojr) [reply src/Common/UserSettingsPersistenceHelper.cs:116]: The new config file gives the users a new clean slate with default settings as they originally used to be. The old config file would for example have `EnableNamingAlias`, `UseDataServiceCollection` and `IgnoreUnexpectedElementsAndAttributes` default values all set with a value of `false` due to previous mishandling of the default values.

File: src/Models/UserSettings.cs (hunk 9 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs:403]: Done

File: src/ODataConnectedServiceWizard.cs (hunk 7 above)
Comment (gathogojr) [reply src/ODataConnectedServiceWizard.cs:70]: The approach I adopting is to have the control binding to a single `UserSettings` instance newed up at [this point](https://github.com/OData/ODataConnectedService/pull/187/files#diff-00999e34e0165406b5257c8c0a958b3a089653a8b8c0cb618d1be46af253f0d3R52). `Endpoint` and `ServiceName` are properties of `UserSettings` and the controls bind to them. That's the reason I'm not adding them to the view model.

File: src/ODataConnectedServiceWizard.cs (hunk 10 above)
Comment (gathogojr) [reply src/ODataConnectedServiceWizard.cs:80]: Whatever code analyzer we're using (or VS default analyzer) considers `this.` superfluous so I dropped it in quite a number of places where possible

File: src/Models/UserSettings.cs (hunk 2 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: Fixed indirectly after refactoring the logic

File: src/Models/UserSettings.cs (hunk 3 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: Yes. We're primarily interested in the MRU endpoints since we always use them to populate the Endpoint dropdown. That way we have everything that we need for purposes of rendering the view. Note that we don't use any other setting from the config file.

File: src/Models/UserSettings.cs (hunk 4 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: Done

File: src/Models/UserSettings.cs (hunk 5 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: Done

File: src/Models/UserSettings.cs (hunk 6 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: That is a concern I also had and set out to address but in the end I had to trade off since using a `ReadOnlyObservableCollection` inhibited two-way binding with the view causing the wizard not to work as expected

File: src/Models/UserSettings.cs [hunk 12]
```diff
@@ -5,147 +5,466 @@
...
+
             if (index >= 0)
             {
-                // Ensure there aren't any duplicates in the list.
-                for (var i = mruList.Count - 1; i > index; i--)
+                // Remove possible duplicates
+                for (var i = this.MruEndpoints.Count - 1; index > 0 && i > index; i--)
                 {
-                    if (EqualityComparer<T>.Default.Equals(mruList[i], item))
+                    if (this.MruEndpoints[i].Equals(mruEndpoint, StringComparison.Ordinal))
```
Comment (KenitoInc) [inline src/Models/UserSettings.cs:448]: We need to be consistent on the use or otherwise of `this.`

File: src/Properties/AssemblyInfo.cs [hunk 13]
```diff
@@ -39,8 +39,8 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("0.11.1.0")]
-[assembly: AssemblyFileVersion("0.11.1.0")]
+[assembly: AssemblyVersion("0.11.2.0")]
```
Comment (KenitoInc) [inline src/Properties/AssemblyInfo.cs:42]: Was the version bump deliberate?
Keep an eye on it during release

**Review by KenitoInc — COMMENTED** [review]
Minor comments

File: src/Properties/AssemblyInfo.cs (hunk 13 above)
Comment (gathogojr) [reply src/Properties/AssemblyInfo.cs:42]: @KenitoInc We'll most definitely need to bump up the version but I'll verify the appropriate version number to use

File: src/Models/UserSettings.cs (hunk 12 above)
Comment (gathogojr) [reply src/Models/UserSettings.cs]: Noted and fixed

**Review by KenitoInc — APPROVED** [review]
LGTM
